From 59610e05a660db59d2a9ea5846648fc2f90ba39d Mon Sep 17 00:00:00 2001
From: Timo Rothenpieler <timo@rothenpieler.org>
Date: Fri, 10 Jan 2025 21:41:55 +0100
Subject: [PATCH 15/21] BACKPORT: avformat/flvdec: add support for legacy HEVC
 files

(cherry picked from commit b76053d8bf322b197a9d07bd27bbdad14fd5bc15)
[Kexy: Resolved minor conflict in libavformat/version.h]
Signed-off-by: Kexy Biscuit <kexybiscuit@aosc.io>
---
 libavformat/flv.h    | 3 +++
 libavformat/flvdec.c | 6 ++++--
 2 files changed, 7 insertions(+), 2 deletions(-)

diff --git a/libavformat/flv.h b/libavformat/flv.h
index 1ea88ff85119..74d3b8de8be9 100644
--- a/libavformat/flv.h
+++ b/libavformat/flv.h
@@ -117,6 +117,9 @@ enum {
     FLV_CODECID_H264    = 7,
     FLV_CODECID_REALH263= 8,
     FLV_CODECID_MPEG4   = 9,
+
+    // non-standard protocol extension that is in use in the wild
+    FLV_CODECID_X_HEVC  = 12,
 };
 
 enum {
diff --git a/libavformat/flvdec.c b/libavformat/flvdec.c
index f167157a0463..e17e5375b242 100644
--- a/libavformat/flvdec.c
+++ b/libavformat/flvdec.c
@@ -380,6 +380,7 @@ static int flv_same_video_codec(AVCodecParameters *vpar, uint32_t flv_codecid)
         return 1;
 
     switch (flv_codecid) {
+    case FLV_CODECID_X_HEVC:
     case MKBETAG('h', 'v', 'c', '1'):
         return vpar->codec_id == AV_CODEC_ID_HEVC;
     case MKBETAG('a', 'v', '0', '1'):
@@ -413,6 +414,7 @@ static int flv_set_video_codec(AVFormatContext *s, AVStream *vstream,
     enum AVCodecID old_codec_id = vstream->codecpar->codec_id;
 
     switch (flv_codecid) {
+    case FLV_CODECID_X_HEVC:
     case MKBETAG('h', 'v', 'c', '1'):
         par->codec_id = AV_CODEC_ID_HEVC;
         vstreami->need_parsing = AVSTREAM_PARSE_HEADERS;
@@ -1658,8 +1660,8 @@ retry_duration:
             }
 
             if (st->codecpar->codec_id == AV_CODEC_ID_MPEG4 ||
-                (st->codecpar->codec_id == AV_CODEC_ID_H264 && (!enhanced_flv || type == PacketTypeCodedFrames)) ||
-                (st->codecpar->codec_id == AV_CODEC_ID_HEVC && type == PacketTypeCodedFrames)) {
+                ((st->codecpar->codec_id == AV_CODEC_ID_H264 || st->codecpar->codec_id == AV_CODEC_ID_HEVC) &&
+                 (!enhanced_flv || type == PacketTypeCodedFrames))) {
                 // sign extension
                 int32_t cts = (avio_rb24(s->pb) + 0xff800000) ^ 0xff800000;
                 pts = av_sat_add64(dts, cts);
-- 
2.48.1.windows.1

